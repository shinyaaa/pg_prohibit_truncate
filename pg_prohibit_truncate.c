#include "postgres.h"

#include "fmgr.h"
#include "tcop/utility.h"

PG_MODULE_MAGIC;

void _PG_init(void);
void _PG_fini(void);

/* Saved hook values in case of unload */
static ProcessUtility_hook_type prev_ProcessUtility = NULL;

static void
prohibit_use_truncate(PlannedStmt *pstmt,
					  const char *queryString,
					  bool readOnlyTree,
					  ProcessUtilityContext context,
					  ParamListInfo params,
					  QueryEnvironment *queryEnv,
					  DestReceiver *dest,
					  QueryCompletion *completionTag)
{
	Node *parsetree = pstmt->utilityStmt;

	if (nodeTag(parsetree) == T_TruncateStmt)
		elog(ERROR, "do not use truncate");
	else
		standard_ProcessUtility(pstmt, queryString,
								readOnlyTree, context,
								params, queryEnv,
								dest, completionTag);
}

void
_PG_init(void)
{
	prev_ProcessUtility = ProcessUtility_hook;
	ProcessUtility_hook = prohibit_use_truncate;
}

void
_PG_fini(void)
{
	ProcessUtility_hook = prev_ProcessUtility;
}