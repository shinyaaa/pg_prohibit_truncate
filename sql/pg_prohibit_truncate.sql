CREATE TABLE test_table (i INT);
INSERT INTO test_table VALUES (1), (2);
SELECT * FROM test_table;
TRUNCATE test_table;
SELECT * FROM test_table;
DROP TABLE test_table;
