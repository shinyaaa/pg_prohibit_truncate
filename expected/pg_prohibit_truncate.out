CREATE TABLE test_table (i INT);
INSERT INTO test_table VALUES (1), (2);
TRUNCATE test_table;
ERROR:  do not use truncate
SELECT * FROM test_table;
 i
---
 1
 2
(2 rows)

DROP TABLE test_table;